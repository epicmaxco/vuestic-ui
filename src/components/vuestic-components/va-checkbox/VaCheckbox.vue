<template>
  <div
    class="va-checkbox"
    :class="computedClass"
  >
    <div
      class="va-checkbox__input-container"
      @click="toggleSelection()"
      @mousedown="hasMouseDown = true"
      @mouseup="hasMouseDown = false"
    >
      <div
        class="va-checkbox__square"
        :class="{'active': isChecked}"
      >
        <input
          :id="id"
          readonly
          @focus="onFocus"
          @blur="isKeyboardFocused = false"
          class="va-checkbox__input"
          @keypress.prevent="toggleSelection()"
          :disabled="disabled"
          :indeterminate="indeterminate"
          :style="inputStyle"
        />
        <va-icon :name="computedIcon"/>
      </div>
      <div
        class="va-checkbox__label-text"
        :style="labelStyle"
      >
        <slot name="label">
          {{ label }}
        </slot>
      </div>
    </div>
    <va-message-list
      class="va-checkbox__error-message-container"
      :value="errorMessages"
      color="danger"
      :limit="errorCount"
    />
  </div>
</template>

<script>
import VaIcon from '../va-icon/VaIcon'
import VaMessageList from '../va-input/VaMessageList'
import { KeyboardOnlyFocusMixin } from './KeyboardOnlyFocusMixin'

export default {
  name: 'va-checkbox',
  components: { VaMessageList, VaIcon },
  mixins: [KeyboardOnlyFocusMixin],
  props: {
    id: String,
    label: String,
    name: String,
    value: {
      type: [Boolean, Array],
      required: true,
    },
    arrayValue: String,
    indeterminate: Boolean,

    disabled: Boolean,
    readonly: Boolean,

    checkedIcon: {
      type: [String, Array],
      default: 'ion ion-md-checkmark',
    },
    indeterminateIcon: {
      type: [String, Array],
      default: 'ion ion-md-remove',
    },

    error: Boolean,
    errorMessages: {
      type: [String, Array],
      default: () => [],
    },
    errorCount: {
      type: Number,
      default: 1,
    },
  },
  computed: {
    computedClass () {
      return {
        'va-checkbox--selected': this.isChecked,
        'va-checkbox--readonly': this.readonly,
        'va-checkbox--disabled': this.disabled,
        'va-checkbox--indeterminate': this.indeterminate,
        'va-checkbox--error': this.showError,
        'va-checkbox--on-keyboard-focus': this.isKeyboardFocused,
      }
    },
    labelStyle () {
      if (this.showError) return { color: this.$themes.danger }
      return {}
    },
    inputStyle () {
      if (this.showError) {
        if (this.isChecked) return { background: this.$themes.danger }
        else return { borderColor: this.$themes.danger }
      } else {
        if (this.isChecked) return { background: this.$themes.primary }
      }
      return {}
    },
    computedIcon () {
      return [
        'va-checkbox__icon-selected',
        this.indeterminate ? this.indeterminateIcon : this.checkedIcon,
      ]
    },
    isChecked () {
      return this.modelIsArray ? this.value.includes(this.arrayValue) : this.value
    },
    modelIsArray () {
      return Array.isArray(this.value)
    },
    showError () {
      // We make error active, if the error-message is not empty and checkbox is not disabled
      if (!this.disabled) {
        if (!(this.errorMessages.length === 0) || this.error) {
          return true
        }
      }
      return false
    },
  },
  methods: {
    toggleSelection () {
      if (this.readonly) {
        return
      }
      if (this.disabled) {
        return
      }
      if (this.modelIsArray) {
        if (this.value.includes(this.arrayValue)) {
          this.$emit('input', this.value.filter(option => option !== this.arrayValue))
        } else {
          this.$emit('input', this.value.concat(this.arrayValue))
        }
        return
      }

      this.$emit('input', !this.value)
    },
  },
}
</script>

<style lang="scss">
@import "../../vuestic-sass/resources/resources";

.va-checkbox {
  display: flex;
  flex-direction: column;
  justify-content: center;

  &__input-container {
    align-items: center;
    display: flex;
    cursor: pointer;

    @at-root {
      .va-checkbox--disabled & {
        @include va-disabled();
      }

      .va-checkbox--readonly & {
        cursor: initial;
      }

      .va-checkbox--disabled & {
        cursor: default;
      }
    }
  }

  #{&}__square {
    @include flex-center();
    width: 2rem;
    height: 2rem;
    position: relative;
    flex: 0 0 2rem;
    @at-root {
      .va-checkbox--on-keyboard-focus#{&} {
        background-color: $light-gray;
        transition: all, 0.6s, ease-in;
        border-radius: 5rem;
      }
    }
  }

  #{&}__input {
    height: 1.375rem;
    width: 1.375rem;
    cursor: inherit;
    color: $white;
    background-color: $white;
    border: solid 0.125rem $gray-light;
    border-radius: 0.25rem;

    &:focus {
      outline: none;
    }

    @at-root {
      .va-checkbox--selected#{&} {
        display: flex;
        justify-content: center;
        align-items: center;
        height: 1.4rem;
        width: 1.4rem;
        color: $white;
        background-color: $vue-green;
        border: 0;
      }

      .va-checkbox--error#{&} {
        border-color: $theme-red;
      }
    }
  }

  #{&}__label-text {
    display: inline-block;
    position: relative;
    margin-left: 0.25rem;
    @at-root {
      .va-checkbox--error#{&} {
        color: $theme-red;
      }
    }
  }

  &__error-message {
    vertical-align: middle;
    color: $theme-red;
    font-size: $font-size-mini;
  }

  &__icon-selected {
    pointer-events: none;
    position: absolute;
    color: $white;
  }

  &__error-message-container {
    flex: 0 0 100%;
    margin-left: 0.3rem; // To fit with checkbox.
  }

  &__label-container {
    margin-left: 2rem;
  }

  &__content {
    flex-direction: row;
  }
}
</style>
