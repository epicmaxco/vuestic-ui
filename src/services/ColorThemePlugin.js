import Vue from 'vue'
import { makeContextablePropsMixin } from '../components/context-test/context-provide/ContextPlugin'

const defaultOptions = Vue.observable({
  themes: {
    primary: '#23e066',
    secondary: '#002c85',
    success: '#40e583',
    info: '#2c82e0',
    danger: '#e34b4a',
    warning: '#ffc200',
    gray: '#b4b6b9',
    dark: '#34495e',
  },
})

export const getDefaultOptions = () => defaultOptions

export const ColorThemePlugin = {
  install (Vue, options) {
    if (options && options.themes) {
      defaultOptions.themes = { ...defaultOptions.themes, ...options.themes }
    }

    Vue.prototype.$themes = defaultOptions.themes
  },
}

const contextConfigMixin = makeContextablePropsMixin({
  color: {
    type: String,
  },
})

export const ColorThemeMixin = {
  mixins: [contextConfigMixin],
  data () {
    return {
      colorThemeDefault: 'primary',
      colorDefault: '#000000',
    }
  },
  computed: {
    _isEnableColorTheme () {
      return Boolean(this.$themes)
    },
    // This allows a multitude of defaults.
    // theme color => color => theme default => hard default
    colorComputed () {
      return this.computeColor(this.c_color)
    },
  },
  methods: {
    computeColor (prop) {
      if (this._isEnableColorTheme) {
        // Use color from global theme config.
        if (this.$themes && this.$themes[prop]) {
          return this.$themes[prop]
        }
      }

      if (prop) {
        // Assume prop is already proper color.
        return prop
      }

      if (this._isEnableColorTheme) {
        // Use theme default
        if (this.$themes && this.$themes[this.colorThemeDefault]) {
          return this.$themes[this.colorThemeDefault]
        }
      }

      // For case when theme is not present and prop is empty.
      return this.colorDefault
    },
  },
}
