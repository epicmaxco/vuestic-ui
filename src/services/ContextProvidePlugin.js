const toCamelCase = (string) => {
  return string.replace(/[_-]/g, ' ').replace(/(?:^\w|[A-Z]|\b\w)/g, (word) => word.toUpperCase()).replace(/\s+/g, '')
}

const mkProvideObject = (options) => {
  const testContext = {
    VaTest: {
      color: 'red',
    },
    VaButton: {
      color: 'purple',
    },
  }

  let _$context = {}

  if (options) {
    _$context = options
  } else {
    _$context = testContext
  }

  return { _$context }
}

export const ContextProvidePlugin = {
  install (Vue, options) {
    Vue.mixin({
      provide: function () {
        return mkProvideObject(options)
      },
    })
  },
}

export const ContextProvideMixin = {
  inject: {
    _$context: {
      from: '_$context',
      default () {
        return this
      },
    },
  },
  methods: {
    _$contextPluginUpdateProps () {
      const tag = this.$options ? this.$options._componentTag : null
      if (tag) {
        const componentContext = this._$context[toCamelCase(tag)]

        Object.assign(this, componentContext)
      }
    },
  },
}
