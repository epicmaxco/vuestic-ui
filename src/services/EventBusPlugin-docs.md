# BusPlugin

This plugin provides mechanism for app-wide communication between the components. This is implementation of publisher-subscriber pattern that is specifically designed to be used inside vue.js components.

To subsribe to events, you should pass the subsription object to the component options:

```js
export default {
  subs: {
    MessageSent (message) {
      this.lastMessage = message
    }
  }
}
```

To fire an event you should call $cast method of your component:

```js
export default {
  methods: {
    fireEvent() {
      this.$cast('MessageSent', this.message)
    }
  }
}
```

$cast may be called from template as well:

```html
<va-button v-on:click="$cast('MessageSent', message)">Fire event from template</va-button>
```