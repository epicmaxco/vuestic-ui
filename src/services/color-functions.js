export const hex2rgb = (hex, opacity) => {
  hex = (hex + '').trim()

  let rgb = null
  let match = hex.match(/^#?(([0-9a-zA-Z]{3}){1,3})$/)

  if (!match) {
    return null
  }

  rgb = {}

  hex = match[1]

  if (hex.length === 6) {
    rgb.r = parseInt(hex.substring(0, 2), 16)
    rgb.g = parseInt(hex.substring(2, 4), 16)
    rgb.b = parseInt(hex.substring(4, 6), 16)
  } else if (hex.length === 3) {
    rgb.r = parseInt(hex.substring(0, 1) + hex.substring(0, 1), 16)
    rgb.g = parseInt(hex.substring(1, 2) + hex.substring(1, 2), 16)
    rgb.b = parseInt(hex.substring(2, 3) + hex.substring(2, 3), 16)
  }

  rgb.css = 'rgb' + (opacity ? 'a' : '') + '('
  rgb.css += rgb.r + ',' + rgb.g + ',' + rgb.b
  rgb.css += (opacity ? ',' + opacity : '') + ')'

  return rgb
}

export const hex2hsl = (H) => {
  /*
   * the source text is taken from here (with minor modifications):
   * https://css-tricks.com/converting-color-spaces-in-javascript/
   */

  // Convert hex to RGB first. Ignore opacity
  let { r, g, b } = hex2rgb(H, 1)

  // Then to HSL
  r /= 255
  g /= 255
  b /= 255

  let cmin = Math.min(r, g, b)
  let cmax = Math.max(r, g, b)
  let delta = cmax - cmin
  let h = 0
  let s = 0
  let l = 0

  if (delta === 0) { h = 0 } else if (cmax === r) { h = ((g - b) / delta) % 6 } else if (cmax === g) { h = (b - r) / delta + 2 } else { h = (r - g) / delta + 4 }

  h = Math.round(h * 60)

  if (h < 0) { h += 360 }

  l = (cmax + cmin) / 2
  s = delta === 0 ? 0 : delta / (1 - Math.abs(2 * l - 1))
  s = +(s * 100).toFixed(1)
  l = +(l * 100).toFixed(1)

  const HSL = {
    h: Math.round(h),
    s: Math.round(s),
    l: Math.round(l),

    get css () {
      return 'hsl(' + HSL.h + ',' + HSL.s + '%,' + HSL.l + '%)'
    },
  }

  return HSL
}

export const getBoxShadowColor = (color) => {
  return hex2rgb(color, 0.6).css
}

export const getHoverColor = (color) => {
  return hex2rgb(color, 0.2).css
}

export const getFocusColor = (color) => {
  return hex2rgb(color, 0.3).css
}

export const getGradientColor = (color) => {
  let first = hex2hsl(color)
  let second = hex2hsl(color)

  // hue circle degrees approximation
  const isRed = (first.h >= 0 && first.h < 45) || (first.h >= 285)
  const isYellow = first.h >= 45 && first.h < 85
  const isGreen = first.h >= 85 && first.h < 165
  const isBlue = first.h >= 165 && first.h < 285
  const isUndersaturated = first.s < 30 // i.e. too pale, gray-ish, monotone

  if (isRed) {
    first.h += 11
    first.s += 27
    first.l += 8
  } else if (isYellow) {
    first.h += 3
    first.l += 9

    second.h -= 2
  } else if (isGreen) {
    first.h += 13
    first.s -= 5
    first.l += 7

    second.h -= 3
    second.s -= 1
    second.l -= 6
  } else if (isBlue) {
    first.h -= 15
    first.s += 3
    first.l += 2
  }

  if (isUndersaturated) {
    first.l += 6
    second.l -= 2
  }

  // restrictions and validations
  if (first.h < 0) { first.h += 360 }
  if (first.h > 0) { Math.round(first.h = first.h % 360) }
  if (second.h < 0) { second.h += 360 }
  if (second.h > 0) { Math.round(second.h = second.h % 360) }

  first.s = first.s > 0 ? first.s : 0
  first.s = first.s < 100 ? first.s : 100
  second.s = second.s > 0 ? second.s : 0
  second.s = second.s < 100 ? second.s : 100

  first.l = first.l > 0 ? first.l : 0
  first.l = first.l < 100 ? first.l : 100
  second.l = second.l > 0 ? second.l : 0
  second.l = second.l < 100 ? second.l : 100

  return [first.css, second.css]
}

export const getGradientBackground = (color) => {
  const [left, right] = getGradientColor(color)
  return `linear-gradient(to right, ${left}, ${right})`
}
