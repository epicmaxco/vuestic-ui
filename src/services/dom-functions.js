export const addOrUpdateStyleElement = (id, getStyles) => {
  if (!id) {
    throw new Error('Style id is required')
  }

  let stylesElement = document.getElementById(id)

  if (stylesElement) {
    stylesElement.innerHTML = getStyles()
  } else {
    stylesElement = document.createElement('style')
    stylesElement.setAttribute('type', 'text/css')
    stylesElement.setAttribute('id', id)
    stylesElement.innerHTML = getStyles()
    document.head.append(stylesElement)
  }
}
