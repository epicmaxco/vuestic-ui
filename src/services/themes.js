export const originalTheme = Object.freeze({
  context: {
    'gradient': true,
    'shadow': 'lg', // 3 states: 'sm', 'lg', undefined (no shadow).
    'invertedColor': false,
  },
  colors: {
    primary: '#40e583',
    secondary: '#002c85',
    success: '#40e583',
    info: '#2c82e0',
    danger: '#e34b4a',
    warning: '#ffc200',
    gray: '#babfc2',
    dark: '#34495e',
  },
})

export const corporateTheme = Object.freeze({
  context: {
    'gradient': false,
    'shadow': 'sm',
    'invertedColor': true,
  },
  colors: {
    primary: '#6c7fee',
    secondary: '#6e7ff1',
    success: '#8ddc88',
    info: '#71baff',
    danger: '#f8706d',
    warning: '#ffd652',
    gray: '#8396a5',
    dark: '#34495e',
  },
})
