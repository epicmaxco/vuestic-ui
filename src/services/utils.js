export const sleep = ms => {
  return new Promise(resolve => setTimeout(resolve, ms))
}

export const normalizeValue = (value, minValue = 0, maxValue = 100) => {
  if (value <= minValue) {
    return minValue
  }

  if (value >= maxValue) {
    return maxValue
  }

  return value
}

// Taken from vuetify pretty much.
export const deepEqual = (a, b) => {
  if (a === b) return true

  // If the values aren't objects, they were already checked for equality
  if (a !== Object(a) || b !== Object(b)) {
    return false
  }

  // Handle dates as a special case
  if (a instanceof Date && b instanceof Date) {
    if (a.getTime() !== b.getTime()) return false
  }

  const props = Object.keys(a)

  // Different number of props, don't bother to check
  if (props.length !== Object.keys(b).length) {
    return false
  }

  return props.every(p => deepEqual(a[p], b[p]))
}

// vuestic-ui warn, currently is mostly a placeholder, but later might provide ability to disable warns etc.
export const warn = (...attrs) => {
  // eslint-disable-next-line  no-console
  console.warn(...attrs)
  return false
}

export const hasOwnProperty = (object, key) => {
  // See https://eslint.org/docs/rules/no-prototype-builtins for explanation
  // on why we don't use object.hasOwnProperty directly.
  return Object.prototype.hasOwnProperty.call(object, key)
}

export const getNestedValue = (option, propsArray) => {
  const lastPropIndex = propsArray.length - 1
  if (lastPropIndex < 0) return option

  let nestedObj = option[propsArray[0]]

  for (let i = 1; i < lastPropIndex; i++) {
    if (nestedObj === null) {
      return option
    }
    nestedObj = nestedObj[propsArray[i]]
  }

  if (nestedObj === null) return option

  return nestedObj[propsArray[lastPropIndex]]
}

export const getValueByPath = (option, prop) => {
  if (option === null || typeof prop !== 'string' || !prop) return option
  if (option[prop] !== undefined) return option[prop]
  prop = prop.replace(/^\./, '') // remove first point symbol
  return getNestedValue(option, prop.split('.'))
}

export const getProp = (option, prop) => {
  if (!prop) return option
  if (typeof prop === 'string') return getValueByPath(option, prop)
  if (typeof prop === 'function') return prop(option)
  return option
}
