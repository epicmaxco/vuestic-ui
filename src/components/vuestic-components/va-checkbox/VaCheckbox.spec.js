import Vue from 'vue'
import { shallowMount } from '@vue/test-utils'
import VaCheckbox from './VaCheckbox'

import { ColorThemePlugin } from '../../../services/ColorThemePlugin'
Vue.use(ColorThemePlugin)

describe('VaCheckbox', () => {
  // TODO: need fix icon with context-config

  it('default', () => {
    const wrapper = shallowMount(VaCheckbox, {
      propsData: { value: false },
    })
    expect(wrapper.html()).toMatchSnapshot()
  })
  it('true value', () => {
    const wrapper = shallowMount(VaCheckbox, {
      propsData: { value: true },
    })
    expect(wrapper.html()).toMatchSnapshot()
  })
})
