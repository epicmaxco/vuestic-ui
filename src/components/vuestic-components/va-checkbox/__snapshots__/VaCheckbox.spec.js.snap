// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VaCheckbox default 1`] = `
<div class="va-checkbox">
  <div class="va-checkbox__input-container">
    <div class="va-checkbox__square"><input id="" readonly="readonly" class="va-checkbox__input">
      <va-icon-stub color="" size="medium" name="check" rotation="" tag="i" class="va-checkbox__icon-selected"></va-icon-stub>
    </div>
    <div class="va-checkbox__label-text">

    </div>
  </div>
  <va-message-list-stub color="danger" value="" limit="1" class="va-checkbox__error-message-container"></va-message-list-stub>
</div>
`;

exports[`VaCheckbox true value 1`] = `
<div class="va-checkbox va-checkbox--selected">
  <div class="va-checkbox__input-container">
    <div class="va-checkbox__square active"><input id="" readonly="readonly" class="va-checkbox__input" style="background: rgb(35, 224, 102);">
      <va-icon-stub color="" size="medium" name="check" rotation="" tag="i" class="va-checkbox__icon-selected"></va-icon-stub>
    </div>
    <div class="va-checkbox__label-text">

    </div>
  </div>
  <va-message-list-stub color="danger" value="" limit="1" class="va-checkbox__error-message-container"></va-message-list-stub>
</div>
`;
